In [898]:
# Imports and pip installations (if needed)
import pandas as pd
import json

# Part 1: Load the dataset

In [899]:
# Load the given datasets
df_num = pd.read_csv('data/chronic_kidney_disease_numerical.csv')
df_cat = pd.read_csv('data/chronic_kidney_disease_categorical.csv')

# Print the data
print(df_num)
print(df_cat)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

The features are the patients (denoted by their patient IDs) and the labels are the numerous variable information columns.

All of the features in the categorical chronic kidney disease dataset are nominal according to the dataset guide, meaning that they are categorical and unordered. All the features in the numerical chronic kidney disease dataset are numerical according to the dataset guide, meaning that they are quantitative and ordered with units.

Three examples of categorical columns are: Pus cell counts (pcc), which can be normal or abnormal. Appetite (appet), which can be good or poor. Anemia (ane) which can be yes or no.

Three examples of numerical columns are: Age (age), which can be an integer representing the patient's age in years. Blood pressure (bp), which is a number in the units of mm/Hg representing the patient's blood pressure. Finally, blood urea (bu) which is a number representing the patient's blood urea levels in mgs/dl.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [900]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicated_rows_num = df_num.duplicated()
if len(duplicated_rows_num) > 0:
    print(duplicated_rows_num)

# Drop these duplicate rows
df_num.drop_duplicates(inplace = True)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicated_rows_cat = df_cat.duplicated()
if len(duplicated_rows_cat) > 0:
    print(duplicated_rows_cat)

df_cat.drop_duplicates(inplace = True)



0      False
1      False
2      False
3      False
4      False
       ...  
415    False
416    False
417    False
418    False
419    False
Length: 420, dtype: bool
0      False
1      False
2      False
3      False
4      False
       ...  
425     True
426     True
427     True
428     True
429     True
Length: 430, dtype: bool


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [901]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df_combined = pd.merge(df_num, df_cat, on = "unique_id")
#Print the combined dataset
print(df_combined)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
389     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [902]:
# Calculate the percentage of rows that contain atleast one missing value
n_missing = df_combined.isnull().any(axis = 1).sum()
print(f"There are {n_missing} rows with at least one missing value")

# Print %
print(f"{round(n_missing * 100 /df_combined.shape[0])}% of rows contain at least one missing value")

# Drop these rows from the dataset
df_combined = df_combined.dropna(how = 'any', axis = 0)

# Print the Dataset
print(df_combined)


There are 237 rows with at least one missing value
61% of rows contain at least one missing value
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0  40.0  1.

## Part 3.4: Sort the dataset according to the Labels

In [903]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df_combined = df_combined.sort_values(by = 'Target')
df_combined = df_combined.reset_index(drop = True)

# Print the dataset
print(df_combined)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [904]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df_combined = pd.get_dummies(df_combined, columns = categorical_columns, dtype = int)

# Print the dataset
print(df_combined)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [905]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for col in numerical_columns:
    # define upper and lower bound
    mean = df_combined[col].mean()
    std = df_combined[col].std()

    upper = mean + 3 * std
    lower = mean - 3 * std

    df_combined = df_combined[(df_combined[col] >= lower) & (df_combined[col] <= upper)]

# Print the dataset
df_combined.reset_index(drop = True, inplace = True)
print(df_combined)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
3       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
4       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
128     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
129     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
130     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
131     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
132     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [906]:
# Normalize the all Numerical Attributes in the dataset.
# I chose to normalize to a range of 0 to 1 to keep things consistent and since it said on campuswire that normalization was up to our interpretation
print(numerical_columns)
for nc in numerical_columns:
    # print(f"range = {df_combined[nc].max() - df_combined[nc].min()}, min = {df_combined[nc].min()}, max = {df_combined[nc].max()}")
    df_combined[nc] = (df_combined[nc] - df_combined[nc].min()) / (df_combined[nc].max() - df_combined[nc].min())

# Print the dataset
print(df_combined)

['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1       137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
2       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
3       532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
4       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
128     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
129     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
130     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
131     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
132     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [907]:
#Remove that column
"""I believe that red blood cell normal and abnormal is not necessary since this information can be derived from the actual
red blood cell count, and it is not that useful since it is highly correlated with red blood cell count."""
to_drop = ['rbc_abnormal', 'rbc_normal']

df_combined = df_combined.drop(columns = to_drop)

# Print the dataset
print(df_combined)

     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1       137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
2       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
3       532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
4       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
128     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
129     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
130     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
131     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
132     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
0    0.206897  0.059406  0.000000  ...  

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [908]:
# Export the dataframe to a new csv file
df_combined.to_csv('clean_data_results.csv')

# Export the dataframe to a new json file
df_combined.to_json('cleaned_data_results.json', orient='records', lines=True)  # orient='records' creates a list of JSON objects


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [909]:
print(df_combined.head(n = 15))

    unique_id       age    bp       bgr        bu        sc       sod  \
0      474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1      137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
2      343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
3      532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
4      546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
5      514721  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
6      621332  0.716216  0.50  1.000000  0.163399  0.111111  0.066667   
7      481293  0.729730  0.00  0.935622  0.169935  0.160494  0.333333   
8      615697  0.000000  0.00  0.103004  0.372549  0.074074  0.500000   
9      843362  0.878378  0.00  0.206009  0.751634  0.604938  0.533333   
10     995177  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
11     828592  0.878378  0.25  0.639485  0.470588  0.395062  0.433333   
12     923613  0.783784  0.00  0.725322  0.313725  

### Paste the markdown table here
| unique_id |      age |    bp |      bgr |       bu |       sc |       sod |       pot |      hemo |       pcv | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----------|----------|-------|----------|----------|----------|-----------|-----------|------------|-----------|-----|--------|---------|------------|-------------|-------|--------|--------|---------|-------------|-----------------|
|     474407| 0.202703 |  0.75 | 0.158798 | 0.196078 | 0.160494 |  0.166667 | 0.206897  | 0.059406   | 0.000000  | ... |      1 |       0 |          1 |           0 |     1 |      0 |      0 |       1 |           1 |               0 |
|     137148| 0.905405 |  1.00 | 0.965665 | 0.522876 | 0.641975 |  0.666667 | 0.000000  | 0.148515   | 0.225806  | ... |      0 |       1 |          0 |           1 |     1 |      0 |      1 |       0 |           1 |               0 |
|     343710| 0.743243 |  0.50 | 0.442060 | 0.901961 | 0.432099 |  0.500000 | 0.793103  | 0.000000   | 0.032258  | ... |      0 |       1 |          0 |           1 |     0 |      1 |      0 |       1 |           1 |               0 |
|     532520| 0.729730 |  0.75 | 0.150215 | 0.281046 | 0.234568 |  0.533333 | 0.793103  | 0.336634   | 0.322581  | ... |      1 |       0 |          1 |           0 |     1 |      0 |      1 |       0 |           1 |               0 |
|     546225| 0.540541 |  0.00 | 0.399142 | 0.535948 | 0.358025 |  0.700000 | 0.379310  | 0.207921   | 0.161290  | ... |      1 |       0 |          1 |           0 |     1 |      0 |      1 |       0 |           1 |               0 |
|     514721| 0.675676 |  0.75 | 0.253219 | 0.633987 | 0.777778 |  0.366667 | 0.655172  | 0.138614   | 0.193548  | ... |      1 |       0 |          1 |           0 |     1 |      0 |      1 |       0 |           1 |               0 |
|     621332| 0.716216 |  0.50 | 1.000000 | 0.163399 | 0.111111 |  0.066667 | 0.206897  | 0.267327   | 0.387097  | ... |      1 |       0 |          0 |           1 |     1 |      0 |      1 |       0 |           1 |               0 |
|     481293| 0.729730 |  0.00 | 0.935622 | 0.169935 | 0.160494 |  0.333333 | 0.034483  | 0.019802   | 0.064516  | ... |      1 |       0 |          0 |           1 |     1 |      0 |      0 |       1 |           1 |               0 |
|     615697| 0.000000 |  0.00 | 0.103004 | 0.372549 | 0.074074 |  0.500000 | 0.689655  | 0.217822   | 0.225806  | ... |      1 |       0 |          0 |           1 |     1 |      0 |      1 |       0 |           1 |               0 |
|     843362| 0.878378 |  0.00 | 0.206009 | 0.751634 | 0.604938 |  0.533333 | 0.689655  | 0.366337   | 0.387097  | ... |      1 |       0 |          0 |           1 |     0 |      1 |      1 |       0 |           1 |               0 |
|     995177| 0.851351 |  0.25 | 0.618026 | 0.562092 | 0.728395 |  0.000000 | 0.344828  | 0.168317   | 0.161290  | ... |      0 |       1 |          1 |           0 |     0 |      1 |      0 |       1 |           1 |               0 |
|     828592| 0.878378 |  0.25 | 0.639485 | 0.470588 | 0.395062 |  0.433333 | 0.517241  | 0.267327   | 0.322581  | ... |      0 |       1 |          1 |           0 |     1 |      0 |      1 |       0 |           1 |               0 |
|     923613| 0.783784 |  0.00 | 0.725322 | 0.313725 | 0.481481 |  0.566667 | 0.862069  | 0.178218   | 0.193548  | ... |      1 |       0 |          0 |           1 |     0 |      1 |      1 |       0 |           1 |               0 |
|     955830| 0.662162 |  0.50 | 0.618026 | 0.411765 | 0.432099 |  0.566667 | 0.689655  | 0.316832   | 0.354839  | ... |      1 |       0 |          1 |           0 |     0 |      1 |      1 |       0 |           1 |               0 |
|     397388| 0.770270 |  1.00 | 0.901288 | 0.163399 | 0.345679 |  0.766667 | 0.206897  | 0.524752   | 0.548387  | ... |      0 |       1 |          1 |           0 |     1 |      0 |      1 |       0 |           1 |               0 |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

I did notice that it hallucinates a bit for datasets. One way to mitigate would be to cross-check LLM-generated outputs using manual validation, automated checks, or by comparing results with traditional methods. You can also split large tasks into smaller chunks and validate each step before proceeding.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [910]:
# Code to rename all the columns in the dataset

# Sample mapping of abbreviations to full names
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns using the mapping
df_combined_renamed = df_combined.rename(columns=column_mapping)

# Display the updated DataFrame
print(df_combined_renamed)


     unique_id       age  blood pressure  blood glucose random  blood urea  \
0       474407  0.202703            0.75              0.158798    0.196078   
1       137148  0.905405            1.00              0.965665    0.522876   
2       343710  0.743243            0.50              0.442060    0.901961   
3       532520  0.729730            0.75              0.150215    0.281046   
4       546225  0.540541            0.00              0.399142    0.535948   
..         ...       ...             ...                   ...         ...   
128     647552  0.743243            0.25              0.214592    0.124183   
129     125587  0.418919            0.00              0.167382    0.241830   
130     614376  0.689189            0.50              0.210300    0.045752   
131     853426  0.472973            0.25              0.236052    0.183007   
132     572140  0.662162            0.50              0.145923    0.117647   

     serum creatinine    sodium  potassium  hemoglobin  packed 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [911]:
# Converted SQL to Pandas code

# Sample DataFrame
data = {
    'Target': ['ckd', 'notckd', 'ckd', 'notckd', 'ckd'],
    # Add other columns as needed
}

df = pd.DataFrame(data)

# Group by 'Target' and count occurrences
result = df.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)


   Target  count
0     ckd      3
1  notckd      2


*Describe what the above code does here*

Overall, this pandas code performs the following actions. First, it creates a DataFrame from a dictionary with a single column named 'Target'. Next, it groups the data by the unique values in the 'Target' column (`'ckd'` and `'notckd'`). After, it counts how many times each unique value appears in the 'Target' column. Then, it resets the index to produce a clean DataFrame with 'Target' values and their corresponding counts. Finally, it prints the resulting counts for each unique 'Target' value. The final result provides a clear statistic about how many occurrences of 'ckd' and 'notckd' there are in the original data.

Both the SQL query and the pandas code achieve the same outcome. They count the occurrences of each unique value in the `Target` column and display those counts. The result will include the unique values in the `Target` column and how many times each appears in the dataset.